### DISCUSSION TITLE
Error when using pypots library

### DISCUSSION THREAD

**Discussion by nguyen-brat** [WenjieDu/PyPOTS#6] [Q&A] [answered]
i install pypots library on my environment successfully but when running the sample code and it give me the error ImportError: cannot import name 'load_specific_dataset' from 'tsdb' (D:\ML\envs\tf_2\lib\site-packages\tsdb\__init__.py). I also try it on google colab but it when i call pip install pypots it running forever, how to fix it thank you !!!!

`import numpy as np`
`from sklearn.preprocessing import StandardScaler`
`from pypots.data import load_specific_dataset, mcar, masked_fill`
`from pypots.imputation import SAITS`
`from pypots.utils.metrics import cal_mae`
`data = load_specific_dataset('physionet_2012')  # For datasets in PyPOTS database, PyPOTS will automatically download and extract it.`
`X = data['X']`
`num_samples = len(X['RecordID'].unique())`
`X = X.drop('RecordID', axis = 1)`
`X = StandardScaler().fit_transform(X.to_numpy())`
`X = X.reshape(num_samples, 48, -1)`
`X_intact, X, missing_mask, indicating_mask = mcar(X, 0.1) # hold out 10% observed values as ground truth`
`X = masked_fill(X, 1 - missing_mask, np.nan)`
`saits = SAITS(n_steps=48, n_features=37, n_layers=2, d_model=256, d_inner=128, n_head=4, d_k=64, d_v=64, dropout=0.1, epochs=10)`
`saits.fit(X)  # train the model. Here I use the whole dataset as the training set, because ground truth is not visible to the model.`
`imputation = saits.impute(X)  # impute the originally-missing values and artificially-missing values`
`mae = cal_mae(imputation, X_intact, indicating_mask)  # calculate mean absolute error on the ground truth (artificially-missing values)`
#
it get the error like this:
ImportError                               Traceback (most recent call last)
Input In [1], in <module>
      1 import numpy as np
      2 from sklearn.preprocessing import StandardScaler
----> 3 from pypots.data import load_specific_dataset, mcar, masked_fill

File D:\ML\envs\tf_2\lib\site-packages\pypots\__init__.py:9, in <module>
      1 """
      2 PyPOTS package.
      3 """
      5 # Created by Wenjie Du <wenjay.du@gmail.com>
      6 # License: GPL-v3
----> 9 from . import data
     10 from . import imputation
     11 from . import utils

File D:\ML\envs\tf_2\lib\site-packages\pypots\data\__init__.py:14, in <module>
      8 from .base import (
      9     BaseDataset,
     10     DatasetForBRITS,
     11     DatasetForMIT,
     12 )
     13 from .generating import generate_random_walk, generate_random_walk_for_classification
---> 14 from .integration import (
     15     fill_nan_with_mask,
     16     mcar,
     17     load_specific_dataset,
     18     list_available_datasets,
     19 )

File D:\ML\envs\tf_2\lib\site-packages\pypots\data\integration.py:10, in <module>
      5 # Created by Wenjie Du <wenjay.du@gmail.com>
      6 # License: GLP-v3
      9 import pycorruptor as corruptor
---> 10 from tsdb import (
     11     load_specific_dataset as _load_specific_dataset,
     12     CACHED_DATASET_DIR as _CACHED_DATASET_DIR,
     13     list_available_datasets as _list_available_datasets,
     14     list_database as _list_database,
     15     list_cached_data as _list_cached_data,
     16     delete_cached_data as _delete_cached_data,
     17     pickle_load as _pickle_load,
     18     pickle_dump as _pickle_dump,
     19 )
     21 CACHED_DATASET_DIR = _CACHED_DATASET_DIR
     22 list_database = _list_database

ImportError: cannot import name 'load_specific_dataset' from 'tsdb' (D:\ML\envs\tf_2\lib\site-packages\tsdb\__init__.py)

**Comment by WenjieDu** [comment, marked as answer]
Hi, which version of TSDB you are using? And PyPOTS.

**Reply by nguyen-brat** [reply]
i using tsdb 0.0.7 and pypots 0.0.4, actually i just call pip install pypots on my environtment and it automatically install for me tsdb and torch

**Reply by WenjieDu** [reply]
Please upgrade PyPOTS to v0.0.7

**Reply by WenjieDu** [reply]
BTW, you can upgrade it with `pip install pypots --upgrade` or `pip install pypots==0.0.7`

**Reply by nguyen-brat** [reply]
i try the fist command but it get error, instead i using pip install --upgrade --no-deps --force-reinstall pypots and seem it work
the error i get is

Building wheel for torch-sparse (setup.py) ... error
ERROR: Command errored out with exit status 1:
command: 'D:\ML\envs\tf_2\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = C:\\Users\\ACER\\AppData\\Local\\Temp\\pip-install-vzdx4cod\\torch-sparse_13f26ee459b841eeb1cbff05f4c4325f\\setup.py'"'"'; __file__='"'"'C:\\Users\\ACER\\AppData\\Local\\Temp\\pip-install-vzdx4cod\\torch-sparse_13f26ee459b841eeb1cbff05f4c4325f\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\ACER\AppData\Local\Temp\pip-wheel-es22w0w2'
       cwd: C:\Users\ACER\AppData\Local\Temp\pip-install-vzdx4cod\torch-sparse_13f26ee459b841eeb1cbff05f4c4325f\
  Complete output (55 lines):
  No CUDA runtime is found, using CUDA_HOME='C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.2'
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.8
  creating build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\add.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\bandwidth.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\cat.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\coalesce.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\convert.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\diag.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\eye.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\index_select.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\masked_select.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\matmul.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\metis.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\mul.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\narrow.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\padding.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\permute.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\reduce.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\rw.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\saint.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\sample.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\select.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\spadd.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\spmm.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\spspmm.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\storage.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\tensor.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\transpose.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\utils.py -> build\lib.win-amd64-3.8\torch_sparse
  copying torch_sparse\__init__.py -> build\lib.win-amd64-3.8\torch_sparse
  running egg_info
  writing torch_sparse.egg-info\PKG-INFO
  writing dependency_links to torch_sparse.egg-info\dependency_links.txt
  writing requirements to torch_sparse.egg-info\requires.txt
  writing top-level names to torch_sparse.egg-info\top_level.txt
  reading manifest file 'torch_sparse.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  warning: no previously-included files matching '*' found under directory 'test'
  adding license file 'LICENSE'
  writing manifest file 'torch_sparse.egg-info\SOURCES.txt'
  running build_ext
  D:\ML\envs\tf_2\lib\site-packages\torch\utils\cpp_extension.py:346: UserWarning: Error checking compiler version for cl: [WinError 2] The system cannot find the file specified
    warnings.warn(f'Error checking compiler version for {compiler}: {error}')
  building 'torch_sparse._convert_cpu' extension
  error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
  Error in atexit._run_exitfuncs:
  Traceback (most recent call last):
    File "D:\ML\envs\tf_2\lib\site-packages\colorama\ansitowin32.py", line 59, in closed
      return stream.closed
  ValueError: underlying buffer has been detached
  ERROR: Failed building wheel for torch-sparse
  Running setup.py clean for torch-sparse
  Building wheel for torch-scatter (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: 'D:\ML\envs\tf_2\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\ACER\\AppData\\Local\\Temp\\pip-install-vzdx4cod\\torch-scatter_6f1e57c8bf4a43a2aa76798363630fe8\\setup.py'"'"'; __file__='"'"'C:\\Users\\ACER\\AppData\\Local\\Temp\\pip-install-vzdx4cod\\torch-scatter_6f1e57c8bf4a43a2aa76798363630fe8\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\ACER\AppData\Local\Temp\pip-wheel-ecukim1_'
       cwd: C:\Users\ACER\AppData\Local\Temp\pip-install-vzdx4cod\torch-scatter_6f1e57c8bf4a43a2aa76798363630fe8\
  Complete output (30 lines):
  No CUDA runtime is found, using CUDA_HOME='C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.2'
  D:\ML\envs\tf_2\lib\site-packages\setuptools\dist.py:717: UserWarning: Usage of dash-separated 'description-file' will not be supported in future versions. Please use the underscore name 'description_file' instead
    warnings.warn(
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.8
  creating build\lib.win-amd64-3.8\torch_scatter
  copying torch_scatter\placeholder.py -> build\lib.win-amd64-3.8\torch_scatter
  copying torch_scatter\scatter.py -> build\lib.win-amd64-3.8\torch_scatter
  copying torch_scatter\segment_coo.py -> build\lib.win-amd64-3.8\torch_scatter
  copying torch_scatter\segment_csr.py -> build\lib.win-amd64-3.8\torch_scatter
  copying torch_scatter\utils.py -> build\lib.win-amd64-3.8\torch_scatter
  copying torch_scatter\__init__.py -> build\lib.win-amd64-3.8\torch_scatter
  creating build\lib.win-amd64-3.8\torch_scatter\composite
  copying torch_scatter\composite\logsumexp.py -> build\lib.win-amd64-3.8\torch_scatter\composite
  copying torch_scatter\composite\softmax.py -> build\lib.win-amd64-3.8\torch_scatter\composite
  copying torch_scatter\composite\std.py -> build\lib.win-amd64-3.8\torch_scatter\composite
  copying torch_scatter\composite\__init__.py -> build\lib.win-amd64-3.8\torch_scatter\composite
  running build_ext
  D:\ML\envs\tf_2\lib\site-packages\torch\utils\cpp_extension.py:346: UserWarning: Error checking compiler version for cl: [WinError 2] The system cannot find the file specified
    warnings.warn(f'Error checking compiler version for {compiler}: {error}')
  building 'torch_scatter._scatter_cpu' extension
  error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
  Error in atexit._run_exitfuncs:
  Traceback (most recent call last):
    File "D:\ML\envs\tf_2\lib\site-packages\colorama\ansitowin32.py", line 59, in closed
      return stream.closed
  ValueError: underlying buffer has been detached

  ERROR: Failed building wheel for torch-scatter
  Running setup.py clean for torch-scatter
Failed to build torch-sparse torch-scatter
Installing collected packages: torch-sparse, torch-scatter, torch-geometric, pypots
    Running setup.py install for torch-sparse ... error
    ERROR: Command errored out with exit status 1:
     command: 'D:\ML\envs\tf_2\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\ACER\\AppData\\Local\\Temp\\pip-install-vzdx4cod\\torch-sparse_13f26ee459b841eeb1cbff05f4c4325f\\setup.py'"'"'; __file__='"'"'C:\\Users\\ACER\\AppData\\Local\\Temp\\pip-install-vzdx4cod\\torch-sparse_13f26ee459b841eeb1cbff05f4c4325f\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\ACER\AppData\Local\Temp\pip-record-a8udj4f5\install-record.txt' --single-version-externally-managed --compile --install-headers 'D:\ML\envs\tf_2\Include\torch-sparse'
         cwd: C:\Users\ACER\AppData\Local\Temp\pip-install-vzdx4cod\torch-sparse_13f26ee459b841eeb1cbff05f4c4325f\
    Complete output (55 lines):
    No CUDA runtime is found, using CUDA_HOME='C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.2'
    running install
    running build
    running build_py
    creating build
    creating build\lib.win-amd64-3.8
    creating build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\add.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\bandwidth.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\cat.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\coalesce.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\convert.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\diag.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\eye.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\index_select.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\masked_select.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\matmul.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\metis.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\mul.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\narrow.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\padding.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\permute.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\reduce.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\rw.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\saint.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\sample.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\select.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\spadd.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\spmm.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\spspmm.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\storage.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\tensor.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\transpose.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\utils.py -> build\lib.win-amd64-3.8\torch_sparse
    copying torch_sparse\__init__.py -> build\lib.win-amd64-3.8\torch_sparse
    running egg_info
    writing torch_sparse.egg-info\PKG-INFO
    writing dependency_links to torch_sparse.egg-info\dependency_links.txt
    writing requirements to torch_sparse.egg-info\requires.txt
    writing top-level names to torch_sparse.egg-info\top_level.txt
    reading manifest file 'torch_sparse.egg-info\SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    warning: no previously-included files matching '*' found under directory 'test'
    adding license file 'LICENSE'
    writing manifest file 'torch_sparse.egg-info\SOURCES.txt'
    running build_ext
    D:\ML\envs\tf_2\lib\site-packages\torch\utils\cpp_extension.py:346: UserWarning: Error checking compiler version for cl: [WinError 2] The system cannot find the file specified
      warnings.warn(f'Error checking compiler version for {compiler}: {error}')
    building 'torch_sparse._convert_cpu' extension
    error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
    Error in atexit._run_exitfuncs:
    Traceback (most recent call last):
      File "D:\ML\envs\tf_2\lib\site-packages\colorama\ansitowin32.py", line 59, in closed
        return stream.closed
    ValueError: underlying buffer has been detached
    ----------------------------------------
ERROR: Command errored out with exit status 1: 'D:\ML\envs\tf_2\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\ACER\\AppData\\Local\\Temp\\pip-install-vzdx4cod\\torch-sparse_13f26ee459b841eeb1cbff05f4c4325f\\setup.py'"'"'; __file__='"'"'C:\\Users\\ACER\\AppData\\Local\\Temp\\pip-install-vzdx4cod\\torch-sparse_13f26ee459b841eeb1cbff05f4c4325f\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\ACER\AppData\Local\Temp\pip-record-a8udj4f5\install-record.txt' --single-version-externally-managed --compile --install-headers 'D:\ML\envs\tf_2\Include\torch-sparse' Check the logs for full command output.

**Reply by WenjieDu** [reply]
This error occurs because, in Windows, torch_sparse has a dependency on Microsoft Visual C++ 14.0 or greater, but it doesn't get satisfied on your machine.

**Reply by nguyen-brat** [reply]
every thing going fine now (at least i can import the library in the sample code and run the sample code on your github) thanks you for your support, but seem like run the pip install --upgrade --no-deps --force-reinstall pypots comment i dont need to update my microsoft visual C++ =)))))

**Reply by WenjieDu** [reply]
Because, with option `--no-deps`, you ignore the dependencies. This is fine with SAITS and other models in PyPOTS. But without torch_sparse, you cannot run model RainDrop.

**Reply by WenjieDu** [reply]
If you use PyPOTS and SAITS and think they are helpful, please star⭐️ them to let more people know about them. This could be your contribution to PyPOTS and SAITS. Thank you in advance.

**Reply by nguyen-brat** [reply]
Sure, my team is using your library and my leader thinks it really helpfull, i will call them star your git

**Reply by WenjieDu** [reply]
Cool, that's really nice of you! Thank you very much for your support and recognition of my work.
Is your team currently working on partially-observed time series? About imputation?

**Reply by nguyen-brat** [reply]
Yeah! my team is woking with a dataset similar to bejjing PM2.5 dataset, some simple imputation method (XGBoost, LightGBM,..) work not really well and we are trying to using another advance and powerfull method !

**Reply by WenjieDu** [reply]
If your dataset is big enough, I believe deep-learning methods can indeed help.

BTW, is partially-observed time series (POTS) common to see in your projects? You know, my research work lies in this field, so I am keen to know about the situation in the industry. This can make me aware of the requirements of the industry, then I can develop models/algorithms accordingly. If you know pain points related to POTS in the industry, please let me know😉 
I'm also working on some new ideas, which will be released in the near future. You can follow me on GitHub, so that you will receive the latest news about my work.

**Comment by nguyen-brat** [comment]
Actually this project is just a competition  I joined with my club so I am
not in the industry of this field and I don't know the demand or pain
related to POTS but maybe in the future, if I know I will mail you later
!!!!!

Vào Th 5, 21 thg 7, 2022 vào lúc 13:21 Wenjie Du ***@***.***>
đã viết:

**Reply by WenjieDu** [reply]
Cool. Many thanks. Your feedback is always welcome ;-)

**Reply by RafaelSee** [reply]
thank you so much mr. wenjie
